File: src/pose/rep_segmentation.py
"""Split a movement clip into repetitions from a single 1-D metric series.

WHY THIS EXISTS. Every `assign_phases` in `src/pose/movements/` was written as if a clip
contained exactly one repetition: a global argmin for the bottom frame, global percentile
thresholds, and hard 15%/85% slices for setup/lockout. On a multi-rep clip that mislabels
every rep after the first. Segmenting first makes those same functions correct, unchanged,
by applying them to one rep at a time.

PORTABILITY CONTRACT (see the RS-SP1 spec §7). RS-SP2 reimplements this in TypeScript so the
browser can decide which spans of a recording to extract densely. Therefore: pure functions,
no I/O, no clock, no global state, and every threshold is a named constant below — never an
inline literal. `tests/fixtures/rep_segmentation_cases.json` pins both implementations to the
same outputs.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence

import numpy as np

# Robust bounds for the signal's dynamic range. Percentiles rather than min/max so one
# mis-detected frame cannot define the range the thresholds are derived from.
PERCENTILE_LOW = 5.0
PERCENTILE_HIGH = 95.0

# Hysteresis band, as fractions of the dynamic range measured from the effort-peak end.
# Two thresholds rather than one: a single threshold would split one rep into several
# whenever the signal wobbles across it near the bottom.
ENTER_FRACTION = 0.35
EXIT_FRACTION = 0.65

# A static-but-jittery signal has a frame-to-frame step of roughly uniform size everywhere,
# so its typical step is a fair estimate of pure noise. A genuine excursion is not uniform: it
# decelerates through its turnarounds (top and bottom), so its CALMEST frame-to-frame steps
# approach the same small size jitter has, even while its fast-moving frames are large. Using
# the median step would conflate that fast-moving portion with noise and, on a fast-cadence
# movement sampled at only a few frames per rep, wrongly call the whole excursion noise. So
# noise is read from the low percentile of steps (the calm ones) via `NOISE_PERCENTILE`, and
# compared against the signal's dynamic range: below this ratio the range is noise, and
# segmenting it would invent repetitions.
NOISE_PERCENTILE = 5.0
MIN_RANGE_TO_NOISE = 6.0

# Movement-agnostic floor on repetition duration. Fast cyclic movements (jumping jacks, high
# knees) legitimately run below this and must lower it — see the spec's §3.4 audit.
DEFAULT_MIN_REP_SECONDS = 0.4

_POLARITIES = ("min", "max")
_REP_STARTS = ("extended", "flexed")


@dataclass(frozen=True)
class RepWindow:
    """One repetition.

    `start`/`end` are inclusive POSITIONS IN THE PASSED SEQUENCE, not `frame_index` values —
    the caller holds the mapping back to frames. `index` is 1-based because it is what a user
    is told ("your 3rd rep").
    """

    index: int
    start: int
    end: int
    partial: bool


def _oriented(signal: Sequence[float], polarity: str, rectify: bool) -> np.ndarray:
    """Normalise any movement's signal to the convention "the effort peak is a LOW value"."""
    values = np.asarray(signal, dtype=np.float64).copy()
    if rectify:
        # A bipolar signal (torso twist: centre -> A -> centre -> B) has two excursions in
        # opposite directions. Rectifying makes each swing its own excursion from zero.
        values = np.abs(values)
    if polarity == "max":
        values = -values
    return values


def _runs_at_or_below(values: np.ndarray, threshold: float) -> list[tuple[int, int]]:
    """Maximal inclusive runs where the signal is at/below `threshold`, skipping NaN."""
    runs: list[tuple[int, int]] = []
    start: int | None = None
    for index, value in enumerate(values):
        if not np.isfinite(value):
            continue  # an unmeasurable frame neither opens nor closes a run
        if value <= threshold and start is None:
            start = index
        elif value > threshold and start is not None:
            runs.append((start, index - 1))
            start = None
    if start is not None:
        runs.append((start, len(values) - 1))
    return runs


def _last_at_or_above(values: np.ndarray, threshold: float, before: int) -> int | None:
    for index in range(before - 1, -1, -1):
        if np.isfinite(values[index]) and values[index] >= threshold:
            return index
    return None


def _first_at_or_above(values: np.ndarray, threshold: float, after: int) -> int | None:
    for index in range(after + 1, len(values)):
        if np.isfinite(values[index]) and values[index] >= threshold:
            return index
    return None


def _climb_backward(values: np.ndarray, index: int) -> int:
    """Walk back from an `exit_` crossing to the top of the excursion.

    The crossing is only where the hysteresis band was pierced; the rep actually begins at the
    standing/extended peak above it. Using the crossing as the boundary would drop the opening
    third of every rep outside the window.
    """
    while index > 0 and np.isfinite(values[index - 1]) and values[index - 1] >= values[index]:
        index -= 1
    return index


def _climb_forward(values: np.ndarray, index: int) -> int:
    """The mirror of `_climb_backward` for the end of a rep."""
    last = len(values) - 1
    while index < last and np.isfinite(values[index + 1]) and values[index + 1] >= values[index]:
        index += 1
    return index


def segment_reps(
    signal: Sequence[float],
    *,
    fps: float,
    polarity: str = "min",
    rectify: bool = False,
    rep_start: str = "extended",
    min_rep_seconds: float = DEFAULT_MIN_REP_SECONDS,
) -> list[RepWindow]:
    """Segment `signal` into repetitions.

    Returns `[]` — never a guess — when the signal carries no repetition structure. The caller
    is required to fall back to whole-clip analysis in that case, NOT to report no faults.
    """
    if polarity not in _POLARITIES:
        raise ValueError(f"polarity must be one of {_POLARITIES}, got {polarity!r}")
    if rep_start not in _REP_STARTS:
        raise ValueError(f"rep_start must be one of {_REP_STARTS}, got {rep_start!r}")

    values = _oriented(signal, polarity, rectify)
    finite = values[np.isfinite(values)]
    min_frames = max(3, int(round(min_rep_seconds * max(fps, 1.0))))
    if finite.size < 2 * min_frames:
        return []

    low = float(np.percentile(finite, PERCENTILE_LOW))
    high = float(np.percentile(finite, PERCENTILE_HIGH))
    span = high - low
    if span <= 0.0:
        return []
    noise = float(np.percentile(np.abs(np.diff(finite)), NOISE_PERCENTILE))
    if noise > 0.0 and span < MIN_RANGE_TO_NOISE * noise:
        return []

    enter = low + ENTER_FRACTION * span
    exit_ = low + EXIT_FRACTION * span
    deep_runs = _runs_at_or_below(values, enter)
    if not deep_runs:
        return []

    if rep_start == "flexed":
        return _windows_from_valleys(values, deep_runs, min_frames)
    return _windows_from_plateaus(values, deep_runs, exit_, min_frames)


def _windows_from_plateaus(
    values: np.ndarray, deep_runs: list[tuple[int, int]], exit_: float, min_frames: int
) -> list[RepWindow]:
    """Boundaries at the EXTENDED end: a rep runs standing -> bottom -> standing.

    Two deep runs with no return above `exit_` between them are one rep, not two — they
    produce the same (start, end) pair here and collapse in the de-duplication below. That
    collapse is the whole point of the hysteresis band.
    """
    spans: list[tuple[int, int, bool]] = []
    for deep_start, deep_end in deep_runs:
        before = _last_at_or_above(values, exit_, deep_start)
        after = _first_at_or_above(values, exit_, deep_end)
        # Cross the band to identify the rep, then climb to the peak to bound it. Two jobs,
        # two mechanisms -- see `_climb_backward`.
        start = 0 if before is None else _climb_backward(values, before)
        end = len(values) - 1 if after is None else _climb_forward(values, after)
        spans.append((start, end, before is None or after is None))
    return _finalize(spans, min_frames)


def _windows_from_valleys(
    values: np.ndarray, deep_runs: list[tuple[int, int]], min_frames: int
) -> list[RepWindow]:
    """Boundaries at the FLEXED end: a rep runs floor -> lockout -> floor (deadlift).

    The span before the first valley and the span after the last are incomplete reps by
    construction, so they are emitted as partial rather than silently dropped.
    """
    valleys: list[int] = []
    for deep_start, deep_end in deep_runs:
        window = values[deep_start : deep_end + 1]
        offset = int(np.nanargmin(np.where(np.isfinite(window), window, np.inf)))
        valleys.append(deep_start + offset)

    spans: list[tuple[int, int, bool]] = []
    if valleys[0] > 0:
        spans.append((0, valleys[0] - 1, True))
    for earlier, later in zip(valleys, valleys[1:]):
        spans.append((earlier, later - 1, False))
    if valleys[-1] < len(values) - 1:
        spans.append((valleys[-1], len(values) - 1, True))
    return _finalize(spans, min_frames)


def _finalize(spans: list[tuple[int, int, bool]], min_frames: int) -> list[RepWindow]:
    """De-duplicate, resolve shared boundaries, drop noise-length spans, and number the rest.

    Adjacent reps meet at a single frame -- the peak between them belongs to the rep that
    STARTS there -- so the earlier window gives it up. Without this, one frame would be
    phase-assigned twice and scored twice.
    """
    unique: list[tuple[int, int, bool]] = []
    seen: set[tuple[int, int]] = set()
    for span in sorted(spans):
        if span[:2] in seen:
            continue
        seen.add(span[:2])
        unique.append(span)

    windows: list[RepWindow] = []
    for position, (start, end, partial) in enumerate(unique):
        if position + 1 < len(unique):
            end = min(end, unique[position + 1][0] - 1)
        if end - start + 1 < min_frames:
            continue
        windows.append(RepWindow(index=len(windows) + 1, start=start, end=end, partial=partial))
    return windows

File: frontend/src/lib/repSegmentation.ts
// Split a movement clip into repetitions from a single 1-D metric series.
//
// A LINE-BY-LINE PORT of src/pose/rep_segmentation.py, which is the authority: read its module
// docstring for why there is deliberately NO noise-vs-range gate (four attempts at one each
// false-rejected ordinary training signal), and why _climbBackward's strict `>` is load-bearing.
// tests/fixtures/rep_segmentation_cases.json pins both implementations to the same outputs.
//
// TWO PORTING TRAPS, both measured, both covered by tests:
//   - numpy's percentile interpolates linearly (np.percentile([1,2,3,4], 5) === 1.15). Picking the
//     k-th element instead shifts the hysteresis band and moves every boundary.
//   - Python's int(round(x)) is banker's rounding. Math.round is not, and the two disagree on
//     6- and 10-rep clips, which would make the languages analyse different repetitions.

/** Robust bounds for the signal's dynamic range, so one bad frame cannot define it. */
export const PERCENTILE_LOW = 5;
export const PERCENTILE_HIGH = 95;
/** Hysteresis band, as fractions of the dynamic range from the effort-peak end. */
export const ENTER_FRACTION = 0.35;
export const EXIT_FRACTION = 0.65;
/** Floor on repetition duration — the ONLY thing separating a real excursion from a blip. */
export const DEFAULT_MIN_REP_SECONDS = 0.4;

const POLARITIES = ["min", "max"] as const;
const REP_STARTS = ["extended", "flexed"] as const;
export type Polarity = (typeof POLARITIES)[number];
export type RepStart = (typeof REP_STARTS)[number];

export interface RepWindow {
  /** 1-based: it is what a user is told ("your 3rd rep"). */
  index: number;
  /** Inclusive POSITION IN THE PASSED SEQUENCE, not a frame_index. */
  start: number;
  end: number;
  partial: boolean;
}

export interface SegmentOptions {
  fps: number;
  polarity?: Polarity;
  rectify?: boolean;
  repStart?: RepStart;
  minRepSeconds?: number;
  /** In ORIENTED space -- i.e. already run through `oriented(signal, polarity, rectify)` below,
   *  not the caller's raw signal. `coarseBand` in repSpans.ts orients internally, using the same
   *  `oriented()` exported from this module, so a `{ polarity, rectify }` matching the ones
   *  passed here is what keeps the two in the same space -- there is no raw-space variant to
   *  misuse. */
  band?: { low: number; high: number };
}

/** numpy.percentile's default linear interpolation, over values that are already sorted. */
export function percentile(sorted: number[], p: number): number {
  if (sorted.length === 1) return sorted[0];
  const position = ((sorted.length - 1) * p) / 100;
  const lower = Math.floor(position);
  const upper = Math.ceil(position);
  if (lower === upper) return sorted[lower];
  return sorted[lower] + (sorted[upper] - sorted[lower]) * (position - lower);
}

/** Python's round(): ties go to the even integer, unlike Math.round which always goes up. */
function roundHalfToEven(value: number): number {
  const floor = Math.floor(value);
  const diff = value - floor;
  if (diff > 0.5) return floor + 1;
  if (diff < 0.5) return floor;
  return floor % 2 === 0 ? floor : floor + 1;
}

/** Normalise any movement's signal to the convention "the effort peak is a LOW value". */
export function oriented(signal: number[], polarity: Polarity, rectify: boolean): number[] {
  return signal.map((value) => {
    // A bipolar signal (torso twist: centre -> A -> centre -> B) has two excursions in opposite
    // directions. Rectifying makes each swing its own excursion from zero.
    const v = rectify ? Math.abs(value) : value;
    return polarity === "max" ? -v : v;
  });
}

/** Maximal inclusive runs at/below `threshold`, skipping non-finite samples. */
function runsAtOrBelow(values: number[], threshold: number): [number, number][] {
  const runs: [number, number][] = [];
  let start: number | null = null;
  values.forEach((value, index) => {
    if (!Number.isFinite(value)) return; // an unmeasurable frame neither opens nor closes a run
    if (value <= threshold && start === null) start = index;
    else if (value > threshold && start !== null) {
      runs.push([start, index - 1]);
      start = null;
    }
  });
  if (start !== null) runs.push([start, values.length - 1]);
  return runs;
}

function lastAtOrAbove(values: number[], threshold: number, before: number): number | null {
  for (let i = before - 1; i >= 0; i -= 1) {
    if (Number.isFinite(values[i]) && values[i] >= threshold) return i;
  }
  return null;
}

function firstAtOrAbove(values: number[], threshold: number, after: number): number | null {
  for (let i = after + 1; i < values.length; i += 1) {
    if (Number.isFinite(values[i]) && values[i] >= threshold) return i;
  }
  return null;
}

/**
 * Walk back from an exit crossing to the top of the excursion, STOPPING AT A PLATEAU.
 *
 * The strict `>` is what makes a window's length equal the excursion's length rather than the
 * clip's, which in turn is what lets `finalize`'s min-frames filter reject a blip on duration
 * alone. See the Python docstring for the full argument.
 */
function climbBackward(values: number[], index: number): number {
  let i = index;
  while (i > 0 && Number.isFinite(values[i - 1]) && values[i - 1] > values[i]) i -= 1;
  return i;
}

function climbForward(values: number[], index: number): number {
  let i = index;
  const last = values.length - 1;
  while (i < last && Number.isFinite(values[i + 1]) && values[i + 1] > values[i]) i += 1;
  return i;
}

/** The full extent of the excursion a deep run belongs to: top, through the bottom, to top. */
function excursionBounds(
  values: number[], deepStart: number, deepEnd: number, exit: number
): [number, number, boolean] {
  const before = lastAtOrAbove(values, exit, deepStart);
  const after = firstAtOrAbove(values, exit, deepEnd);
  const start = before === null ? 0 : climbBackward(values, before);
  const end = after === null ? values.length - 1 : climbForward(values, after);
  return [start, end, before === null || after === null];
}

/** Boundaries at the EXTENDED end: a rep runs standing -> bottom -> standing. */
function windowsFromPlateaus(
  values: number[], deepRuns: [number, number][], exit: number, minFrames: number
): RepWindow[] {
  return finalize(deepRuns.map(([s, e]) => excursionBounds(values, s, e, exit)), minFrames);
}

/**
 * Boundaries at the FLEXED end: a rep runs floor -> lockout -> floor (deadlift).
 *
 * Filters each deep run on the duration of the excursion it belongs to, because a valley-to-valley
 * window's length is the rep PERIOD, not any one excursion — without this the flexed path has no
 * anomaly rejection at all. See the Python docstring for the boundary-run trade this accepts.
 */
function windowsFromValleys(
  values: number[], deepRuns: [number, number][], exit: number, minFrames: number
): RepWindow[] {
  const realRuns = deepRuns.filter(([s, e]) => {
    const [start, end] = excursionBounds(values, s, e, exit);
    return end - start + 1 >= minFrames;
  });
  if (realRuns.length === 0) return [];

  const valleys = realRuns.map(([s, e]) => {
    let best = s;
    let bestValue = Infinity;
    for (let i = s; i <= e; i += 1) {
      if (Number.isFinite(values[i]) && values[i] < bestValue) { bestValue = values[i]; best = i; }
    }
    return best;
  });

  const spans: [number, number, boolean][] = [];
  if (valleys[0] > 0) spans.push([0, valleys[0] - 1, true]);
  for (let i = 0; i + 1 < valleys.length; i += 1) spans.push([valleys[i], valleys[i + 1] - 1, false]);
  if (valleys[valleys.length - 1] < values.length - 1) {
    spans.push([valleys[valleys.length - 1], values.length - 1, true]);
  }
  return finalize(spans, minFrames);
}

/** De-duplicate, resolve shared boundaries, drop too-short spans, and number the rest. */
function finalize(spans: [number, number, boolean][], minFrames: number): RepWindow[] {
  const sorted = [...spans].sort((a, b) =>
    a[0] - b[0] || a[1] - b[1] || Number(a[2]) - Number(b[2]));

  const unique: [number, number, boolean][] = [];
  const seen = new Set<string>();
  for (const span of sorted) {
    const key = `${span[0]}:${span[1]}`;
    if (seen.has(key)) continue;
    seen.add(key);
    unique.push(span);
  }

  const windows: RepWindow[] = [];
  unique.forEach(([start, rawEnd, partial], position) => {
    // Adjacent reps meet at a single frame — the peak between them belongs to the rep that STARTS
    // there — so the earlier window gives it up, or one frame is phased and scored twice.
    const end = position + 1 < unique.length ? Math.min(rawEnd, unique[position + 1][0] - 1) : rawEnd;
    if (end - start + 1 < minFrames) return;
    windows.push({ index: windows.length + 1, start, end, partial });
  });
  return windows;
}

/**
 * Segment `signal` into repetitions.
 *
 * Returns `[]` — never a guess — when the signal carries no repetition structure. The caller is
 * required to fall back to whole-clip analysis in that case, NOT to report no faults.
 */
export function segmentReps(signal: number[], options: SegmentOptions): RepWindow[] {
  const polarity = options.polarity ?? "min";
  const repStart = options.repStart ?? "extended";
  const minRepSeconds = options.minRepSeconds ?? DEFAULT_MIN_REP_SECONDS;
  if (!POLARITIES.includes(polarity)) throw new Error(`polarity must be min or max, got ${polarity}`);
  if (!REP_STARTS.includes(repStart)) throw new Error(`repStart must be extended or flexed, got ${repStart}`);

  const values = oriented(signal, polarity, options.rectify ?? false);
  const finite = values.filter(Number.isFinite).sort((a, b) => a - b);
  const minFrames = Math.max(3, roundHalfToEven(minRepSeconds * Math.max(options.fps, 1)));
  if (finite.length < 2 * minFrames) return [];

  // The band may be supplied by the caller. RS-SP2 refines a rep's boundary inside a padded span,
  // and a span's OWN percentiles are computed over one repetition's worth of samples -- narrow
  // enough that the hysteresis band shifts and the boundary moves with it. Measured on 46 real
  // clips: per-span percentiles refine 92.9% of reps exactly (p95 15.3 frames, max 46), while the
  // same spans given the whole clip's range refine 98.6% exactly (p95 0, max 1). The caller has a
  // whole-clip range available because the coarse pass covers the whole clip. See the SP2 spec
  // §2.1.1, and `coarseBand` in repSpans.ts.
  const low = options.band ? options.band.low : percentile(finite, PERCENTILE_LOW);
  const high = options.band ? options.band.high : percentile(finite, PERCENTILE_HIGH);
  const span = high - low;
  if (span <= 0) return [];

  const enter = low + ENTER_FRACTION * span;
  const exit = low + EXIT_FRACTION * span;
  const deepRuns = runsAtOrBelow(values, enter);
  if (deepRuns.length === 0) return [];

  return repStart === "flexed"
    ? windowsFromValleys(values, deepRuns, exit, minFrames)
    : windowsFromPlateaus(values, deepRuns, exit, minFrames);
}

/**
 * Choose which repetitions to actually analyze: first / middle / last.
 *
 * Not "the first N": the first rep carries warm-up errors, the middle one steady state, the last
 * one fatigue breakdown, and sampling only the middle systematically hides the fault a lifter most
 * needs told. Partial reps are skipped when complete ones exist, but kept when they are all there
 * is — analyzing a truncated rep beats analyzing nothing.
 */
export function selectReps(reps: RepWindow[], maxReps: number | null): RepWindow[] {
  const complete = reps.filter((rep) => !rep.partial);
  const candidates = complete.length > 0 ? complete : [...reps];
  if (candidates.length === 0) return [];
  if (!maxReps || maxReps <= 0 || candidates.length <= maxReps) return candidates;

  // np.linspace(start, stop, num) with num === 1 returns [start], not a division by (num - 1);
  // that division only makes sense for spacing out 2+ points. Guarding it here mirrors that
  // numpy behaviour directly instead of letting `(last * i) / (maxReps - 1)` divide by zero at
  // i === 0, which would produce NaN and silently smuggle `candidates[NaN] === undefined` into
  // the result. Python's `select_reps(reps, max_reps=1)` returns `[candidates[0]]`.
  if (maxReps === 1) return [candidates[0]];

  const last = candidates.length - 1;
  const positions = new Set<number>();
  for (let i = 0; i < maxReps; i += 1) {
    positions.add(roundHalfToEven((last * i) / (maxReps - 1)));
  }
  return [...positions].sort((a, b) => a - b).map((position) => candidates[position]);
}
